### PR TITLE
Writer lock reason

### PR DESCRIPTION
**Pull request by patrickSpaceSurfer**
Add a reason string when locking a table with the writer

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 66 / 73 (90.41%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|6|9|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|15|18|83.33%|
|:large_blue_circle:|io/questdb/cairo/pool/ex/EntryLockedException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/EntryUnavailableException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/InsertStatementImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/TelemetryJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/CairoLineProtoParser.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 66 / 73 (90.41%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|6|9|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|15|18|83.33%|
|:large_blue_circle:|io/questdb/cairo/pool/ex/EntryLockedException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/EntryUnavailableException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/InsertStatementImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/TelemetryJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/CairoLineProtoParser.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 66 / 73 (90.41%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|6|9|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|15|18|83.33%|
|:large_blue_circle:|io/questdb/cairo/pool/ex/EntryLockedException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/EntryUnavailableException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/InsertStatementImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/TelemetryJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/CairoLineProtoParser.java|2|2|100.00%|

File: core/src/main/java/io/questdb/cairo/pool/WriterPool.java [hunk 1]
```diff
@@ -197,12 +196,17 @@ public boolean lock(CharSequence tableName) {
         // try to change owner
         if ((Unsafe.cas(e, ENTRY_OWNER, UNALLOCATED, thread) /*|| Unsafe.cas(e, ENTRY_OWNER, thread, thread)*/)) {
             closeWriter(thread, e, PoolListener.EV_LOCK_CLOSE, PoolConstants.CR_NAME_LOCK);
-            return lockAndNotify(thread, e, tableName);
+            if (lockAndNotify(thread, e, tableName, lockReason)) {
+                return null;
+            }
+            assert null != e.ownershipReason;
+            return e.ownershipReason;
```
Comment (ideoma) [inline core/src/main/java/io/questdb/cairo/pool/WriterPool.java:203]: I'm paranoiac and  despite the asserts would change to
```
return e.ownershipReason != null ? e.ownershipReason : "UNKNOWN";
```

**Review by bluestreak01 — CHANGES_REQUESTED** [review]
Lock reason is not reported:

`drop table`: 

image

`truncate table`:

image

**Comment by bluestreak01** [comment]
`rename` of locked table has become possible (regressed), without any test failing, which is a concern

File: core/src/main/java/io/questdb/cairo/pool/WriterPool.java (hunk 1 above)
Comment (patrickSpaceSurfer) [reply core/src/main/java/io/questdb/cairo/pool/WriterPool.java:203]: Well we have very high test coverage around this area. What might be a good idea is to through a null pointer exception if someone tries to set a null reason

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 66 / 75 (88.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|6|9|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|15|18|83.33%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|24|26|92.31%|
|:large_blue_circle:|io/questdb/cairo/pool/ex/EntryLockedException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/EntryUnavailableException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/InsertStatementImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/TelemetryJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/CairoLineProtoParser.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 77 / 87 (88.51%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|6|9|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|15|18|83.33%|
|:large_blue_circle:|io/questdb/tasks/TelemetryTask.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|24|26|92.31%|
|:large_blue_circle:|io/questdb/cairo/pool/ex/EntryLockedException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/EntryUnavailableException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/InsertStatementImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/TelemetryJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/CairoLineProtoParser.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 77 / 87 (88.51%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|6|9|66.67%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|15|18|83.33%|
|:large_blue_circle:|io/questdb/tasks/TelemetryTask.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|24|26|92.31%|
|:large_blue_circle:|io/questdb/cairo/pool/ex/EntryLockedException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/EntryUnavailableException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/InsertStatementImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/TelemetryJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/CairoLineProtoParser.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 83 / 93 (89.25%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|7|10|70.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/tasks/TelemetryTask.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|27|29|93.10%|
|:large_blue_circle:|io/questdb/cairo/pool/ex/EntryLockedException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/EntryUnavailableException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/InsertStatementImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/TelemetryJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/CairoLineProtoParser.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 83 / 93 (89.25%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|7|10|70.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/tasks/TelemetryTask.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|27|29|93.10%|
|:large_blue_circle:|io/questdb/cairo/pool/ex/EntryLockedException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/EntryUnavailableException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/InsertStatementImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/TelemetryJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/CairoLineProtoParser.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 83 / 93 (89.25%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|7|10|70.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/tasks/TelemetryTask.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|27|29|93.10%|
|:large_blue_circle:|io/questdb/cairo/pool/ex/EntryLockedException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/EntryUnavailableException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/InsertStatementImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/TelemetryJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/CairoLineProtoParser.java|2|2|100.00%|

**Comment by patrickSpaceSurfer** [comment]
> Lock reason is not reported:
> 
> `drop table`:
> 
> image
> 
> `truncate table`:
> 
> image

I have updated the error string for these cases

**Comment by patrickSpaceSurfer** [comment]
> `rename` of locked table has become possible (regressed), without any test failing, which is a concern

I cant reproduce this

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 83 / 93 (89.25%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/pgwire/PGConnectionContext.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|7|10|70.00%|
|:large_blue_circle:|io/questdb/cairo/CairoEngine.java|17|20|85.00%|
|:large_blue_circle:|io/questdb/tasks/TelemetryTask.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/pool/WriterPool.java|27|29|93.10%|
|:large_blue_circle:|io/questdb/cairo/pool/ex/EntryLockedException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cutlass/text/CairoTextWriter.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/EntryUnavailableException.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/InsertStatementImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/TelemetryJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/pool/ReaderPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/udp/CairoLineProtoParser.java|2|2|100.00%|

**Pull request merged**